### PR TITLE
Fixes #9021: Patch the agent to output needed data for jinja2 templating

### PR DESCRIPTION
**Pull request by amousset**
https://www.rudder-project.org/redmine/issues/9021

### CODE REVIEW

**Comment by amousset** [comment]
Still testing, please wait before merging.

File: rudder-agent/SOURCES/patches/cfengine/80-add-dumpdatatstate-function.patch [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
++
++
+ static FnCallResult FnCallBundlestate(EvalContext *ctx,
+                                       ARG_UNUSED const Policy *policy,
+                                       ARG_UNUSED const FnCall *fp,
+@@ -8109,6 +8135,8 @@ const FnCallType CF_FNCALL_TYPES[] =
+                   FNCALL_OPTION_NONE, FNCALL_CATEGORY_IO, SYNTAX_STATUS_NORMAL),
+     FnCallTypeNew("datastate", CF_DATA_TYPE_CONTAINER, DATASTATE_ARGS, &FnCallDatastate, "Construct a container of the variable and class state",
+                   FNCALL_OPTION_NONE, FNCALL_CATEGORY_UTILS, SYNTAX_STATUS_NORMAL),
++    FnCallTypeNew("dumpdatastate", CF_DATA_TYPE_CONTEXT, FILESTAT_ARGS, &FnCallDumpDatastate, "Dump the current datatstae() into the given file",
```
Comment (jooooooon) [inline rudder-agent/SOURCES/patches/cfengine/80-add-dumpdatatstate-function.patch]: There's a typo here: datastate()*

File: rudder-agent/SOURCES/patches/cfengine/80-add-dumpdatatstate-function.patch [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
++                                        ARG_UNUSED const Policy *policy,
++                                        ARG_UNUSED const FnCall *fp,
++                                        ARG_UNUSED const Rlist *args)
++{
++	    char *path = RlistScalarValue(args);
++
++	    FILE *fout;
++            if ((fout = fopen(path, "w")) == NULL)
++	    {
++	       Log(LOG_LEVEL_ERR, "Cannot open the destination file '%s'. (fopen: %s)",
```
Comment (jooooooon) [inline rudder-agent/SOURCES/patches/cfengine/80-add-dumpdatatstate-function.patch]: It's useful in error messages to know where this error is coming from. Could you prefix it with "dumpdatastate(): " maybe?

File: rudder-agent/SOURCES/patches/cfengine/80-add-dumpdatatstate-function.patch [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+@@ -2956,6 +2956,32 @@ static FnCallResult FnCallDatastate(Eval
+     return  (FnCallResult) { FNCALL_SUCCESS, (Rval) { state, RVAL_TYPE_CONTAINER } };
+ }
+ 
++static FnCallResult FnCallDumpDatastate(EvalContext *ctx,
++                                        ARG_UNUSED const Policy *policy,
++                                        ARG_UNUSED const FnCall *fp,
++                                        ARG_UNUSED const Rlist *args)
++{
++	    char *path = RlistScalarValue(args);
```
Comment (jooooooon) [inline rudder-agent/SOURCES/patches/cfengine/80-add-dumpdatatstate-function.patch]: Doesn't this need free-ing at the end of the function?

File: rudder-agent/SOURCES/patches/cfengine/80-add-dumpdatatstate-function.patch [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
++	           path, GetErrorStr());
++	       return FnReturnContext(false);
++	    }
++
++            Writer *writer = FileWriter(fout);
++
++            JsonElement *state = DefaultTemplateData(ctx, NULL);
++            JsonWrite(writer, state, 0);
++
++            WriterClose(writer);
```
Comment (jooooooon) [inline rudder-agent/SOURCES/patches/cfengine/80-add-dumpdatatstate-function.patch]: Does this also fclose() the FILE handle above? If not we should close it.

File: rudder-agent/SOURCES/patches/cfengine/80-add-dumpdatatstate-function.patch (hunk 3 above)
Comment (amousset) [inline rudder-agent/SOURCES/patches/cfengine/80-add-dumpdatatstate-function.patch]: No, the data is in args, which is freed by FnCallDestroy in a parent function.

File: rudder-agent/SOURCES/patches/cfengine/80-add-dumpdatatstate-function.patch (hunk 2 above)
Comment (amousset) [inline rudder-agent/SOURCES/patches/cfengine/80-add-dumpdatatstate-function.patch]: sure!

File: rudder-agent/SOURCES/patches/cfengine/80-add-dumpdatatstate-function.patch (hunk 4 above)
Comment (amousset) [inline rudder-agent/SOURCES/patches/cfengine/80-add-dumpdatatstate-function.patch]: sure!

**Comment by amousset** [comment]
Commit modified

**Comment by peckpeck** [comment]
Looks good to me

**Pull request merged**